src/dbnode/storage/series/buffer.go

```diff
@@ -281,6 +311,26 @@ func (b *dbBuffer) IsEmpty() bool {
 	return len(b.bucketsMap) == 0
 }
 
+func (b *dbBuffer) ColdFlushBlockStarts(latestWarmFlushTime time.Time) OptimizedTimes {
+	var times OptimizedTimes
+
+	for t, bucketVersions := range b.bucketsMap {
+		for _, bucket := range bucketVersions.buckets {
+			if bucket.version == writableBucketVersion &&
```

Comment:
So basically this is checking to see if there is a writableBucket to determine that this blockStart has data that needs to be cold flushed? What about the following scenario:

1. Block X has cold data that needs to be flushed
2. ColdFlush comes in and "flushes" the data, incrementing the bucketVersion, but the cold flush ultimately fails and the data is not written to disk. No subsequent writes come in for this bucket.
3. Next cold flush comes in and calls ColdFlushBlockStarts and thinks this block doesn't have any data that needs to be cold flushed because the version was already incremented

Does this function need to accept the current successful cold flush version, similar to the tick function?

Reply:
Nice catch on this edge case. Since this function goes through all its block starts, I'll pass through a block state snapshot, which is the same as what we do for the Tick.